[로그인 및 access token 재발급] 로그인 및 access token 재발급 비즈니스 로직/API 구성(#50)

Minjae-An opened the pull request:
## 작업 대상
<!-- 작업한 대상을 자세히 설명해주세요. -->
- 로그인 비즈니스 로직 및 API
- access token refresh 비즈니스 로직 및 API
- 인가 로직
- 화살 내역 엔티티

## 📄 작업 내용
<!-- 작업한 내용을 간단히 요약해주세요. -->
- 로그인 비즈니스 로직 및 API 구현
- refresh token 바탕 access token을 재발급 받는 비즈니스 로직 및 API 구현
- access token으로 사용자에게 인가 과정을 수행하는 `BearerTokenAuthInterceptor` 구현
- 위 작업 내용 연관 로직 수정
- 화살 내역 엔티티 필드명 변경에 따른 관련 코드 수정

## 🙋🏻 주의 사항
<!-- 리뷰어를 위해 복잡하거나 중요한 코드를 명시해주세요. -->
- 클라이언트가 `refresh token`을 쿠키에 포함하여 access token 재발급 요청시 사용하도록 
`refresh token` 발급 로직 구성
- 서버에서 쿠키를 통해  `refresh token`을 전달 받도록 발급 로직을 구성하며 다음 사항들을 고려
   - HTTPS 환경에서만 해당 쿠키를 요청에 포함하도록 `secure=true` 속성 설정
   - XSS 공격 방지를 위해 `HttpOnly=true` 속성을 설정하여 JS 코드에서 접근이 불가하게 설정
   - CSRF 공격 방지를 위해 `SameSite=Strict` 속성 설정
   - access token 재발급 시에만 해당 쿠키를 요청에 포함토록 `path=/users/refresh` 설정
- 회원 가입 완료시 access token과 refresh token을 사용자에게 제공, 앞서 구성한 refresh token 발급 로직 사용
- access token을 response body를 통해 제공할 때 `Bearer` 토큰 형식으로 전달토록 토큰 생성 로직 구성
- `LoginService`에서 `LoginController`로 로그인 로직 수행 결과를 제공할 때 `UserLoginResult`라는 별도의 객체를
  사용하여 필요한 데이터를 전달받도록 구성
  - HTTP response를 조작하는 코드는 클래스 역할상 controller에 위치해야 한다고 판단
  - `LoginService`에서 필요한 데이터를 `LoginController`에 제공하고 controller가 response header/body에 
    포함되어야 할 데이터를 적절히 가공하여 클라이언트에 전달하는 구조로 로직 구성

## 📎 관련 이슈
- 출석 체크 비즈니스 로직 및 API 관련 이슈는 별도 브랜치를 통해 작업 예정

## 레퍼런스
<!-- 작업 시 참고했던 문건의 URL을 남겨주세요 -->
- https://velog.io/@iamtaehoon/sagah#%EB%A7%88%EC%B9%98%EB%A9%B0
- https://minjung-jeon.github.io/same-site-same-origin/

Minjae-An committed:
feat: 화살 보내기 요청 DTO 추가(#24)

보낼 화살 수를 담은 요청 DTO 정의

Minjae-An committed:
feat: 확살 송수신 내역 repository, 내역 확인 로직 추가(#24)

- 화살 송수신 내역 repository 정의
- 보낸 사용자, 받은 사용자를 통해 내역 존재 여부를 확인하는 로직 구현

Minjae-An committed:
feat: 화살 송수신 내역 존재 여부 확인 로직 추가(#24)

보낸 사용자, 받은 사용자를 통해 송수신 내역 존재 여부를 확인하는
로직 구현

Minjae-An committed:
feat: 화살 송수신 내역 저장 기능 구현(#24)

Minjae-An committed:
feat: 화살 관련 예외 enum 정의(#24)

다음 상황들에 대한 예외 enum을 정의하였다.
- 자기 자신에게 화살을 보내는 경우
- 이미 화살을 보낸 사용자인 경우
- 가진 화살 수가 부족해 화살을 보낼 수 없는 경우

Minjae-An committed:
feat: 사용자 화살 감소 로직 추가(#24)

감소하려는 화살 수가 사용자가 보유한 화살 수보다 클 경우 예외 발생

Minjae-An committed:
refactor: 생성 시간 필드, 불필요한 업데이트 불가 DDL 속성 삭제(#24)

Minjae-An committed:
feat: 화살 보내기 비즈니스 로직 추가(#24)

화살 보내기 비즈니스 로직은 다음 과정을 거친다.
1. 자기 자신에게 화살을 보내는 상황 검증
2. 화살을 보낸 사용자에게 또 보내는 상황 검증
3. 화살 내역 저장
4. 보내는 사용자 화살 감소, 화살이 부족할 경우 예외 발생(과정 3 롤백)
5. 받는 사용자 화살 증가

Minjae-An committed:
feat: 화살 보내기 API 구현(#24)

Minjae-An committed:
refactor: 화살 command repository 삭제(#24)

- query, command를 service, repository를 중개하는 별도의 계층을 통해 구분
- 이에 따라 기존 command repository 삭제
- 기존 command repository 사용 위치, 화살 command로 대체

Minjae-An committed:
refactor: 코드 정렬 수정(#24)

Minjae-An committed:
Merge branch 'dev' into feat/#24-send-arrow

# Conflicts:
#	be/src/main/java/yeonba/be/arrow/repository/ArrowCommand.java
#	be/src/main/java/yeonba/be/arrow/service/ArrowService.java
#	be/src/main/java/yeonba/be/exception/CommonException.java
#	be/src/main/java/yeonba/be/user/entity/User.java

Minjae-An committed:
refactor: 사용자 화살 증감 메서드 이름 수정(#24)

서로 대치되는 작업을 수행하는 상황을 잘 표현할 수 있도록 메서드 이름 수정

Minjae-An committed:
feat: 화살 관련 예외 enum 추가(#24)

- 이미 화살을 보낸 사용자 예외
- 화살이 부족하여 화살을 보낼 수 없는 예외

Minjae-An committed:
refactor: 화살 내역 존재 여부 확인 메서드 이름 수정(#24)

가독성 향상을 위해 간단한 이름을 수정

Minjae-An committed:
refactor: 코드 배치 수정(#24)

dev 브랜치 작업 내역 병합에 따른 코드 배치 수정

Minjae-An committed:
refactor: 화살 보내기 API 파라미터 이름 수정(#24)

좀 더 명확하게 대치되는 의미를 나타낼 수 있도록 화살을 받는
사용자 ID 파라미터 이름을 receiverId로 수정

Minjae-An committed:
refactor: 컨벤션에 맞게 코드 배치 수정(#24)

Minjae-An committed:
feat: 이름 칼럼, 생성자 추가(#27)

- 이름 칼럼 추가
- 테스트시 활용 가능한 생성자 추가

Minjae-An committed:
feat: 상대 프로필 응답 DTO 수정(#27)

- 래퍼 타입 필드들 기본 타입으로 변경
- 프로필 사진 URL 리스트 필드 추가
- 사용자가 가진 총 화살 수 필드 추가
- 음주 성향, 흡연 성향 필드 추가
- 코드 정렬

Minjae-An committed:
feat: 사용자 엔티티 수정(#27)

- 레퍼런스 타입 not null 필드들, Column 어노테이션 사용 명시
- salt 필드 추가
- 생성 일시, 최종 수정 일시 필드 추가
- JpaAuditing 활용, 생성/최종 수정 일시 필드들 엔티티 저장시 자동 설정
- 새로운 생성자 구성

Minjae-An committed:
feat: 상대 프로필 조회 비즈니스 로직 추가(#27)

프로필을 조회하는 사용자 ID의 경우 상대방이 화살을 보넀던 사용자인지
확인하기 위해 파라미터로 받는다.

Minjae-An committed:
feat: 상대 프로필 API 구현 및 코드 정렬(#27)

Minjae-An committed:
feat: 사용자 엔티티 수정(#27)

- 요구사항 변경에 따라 음주 성향, 흡연 성향 필드 삭제
- 성별 정보 제공 메서드 추가

Minjae-An committed:
feat: 사용자 프로필 응답 DTO 수정(#27)

- 음주 성향, 흡연 성향 필드 삭제
- 성별 필드 추가

Minjae-An committed:
feat: 사용자 프로필 상세 응답 DTO 수정(#27)

- 음주 성향, 흡연 성향 필드 삭제
- 생성자 성별 필드 설정 로직 수정
- 코드 정렬

Minjae-An committed:
feat: 사용자 프로필 조회 비즈니스 로직 수정(#27)

- 음주 성향, 흡연 성향 삭제
- 성별 정보 응답에 포함토록 수정
- 코드 정렬

Minjae-An committed:
chore: spring mail 종속성 추가(#35)

Minjae-An committed:
chore: Gmail SMTP 사용을 위한 설정 추가(#35)

Minjae-An committed:
feat: 이메일 기반 사용자 조회 로직 추가(#35)

Minjae-An committed:
feat: 이메일 기반 사용자 조회 로직 추가(#35)

Minjae-An committed:
feat: 서비스 사용 정규식 enum 정의(#35)

- 서비스에서 사용하는 정규식들을 모아놓는 enum 클래스 정의
- 이메일, 비밀번호 정규식 정의

Minjae-An committed:
feat: 임시 비밀번호 발급 기능 구현(#35)

- 비밀번호 정규식을 충족하는 임의의 임시 비밀번호를 발급
- Random보다 생성하는 난수를 예측하기 어려워 보안적으로 뛰어난 SecureRandom 사용

Minjae-An committed:
feat: 메일 관련 설정 정의(#35)

- spring mail에서 이메일 전송을 위해 제공하는 JavaMailSender 빈 등록
- 필요한 설정 값들은 application.yml에 정의된 값들을 끌어와 사용
- 빈 등록 로직에서 구글 SMTP 서버를 통해 메일을 전송하기 위한 설정 코드 추가

Minjae-An committed:
feat: 이메일 전송 로직 추가(#35)

- 단순 텍스트 이메일 전송 기능 구현
- 발신자는 application.yml에 정의해논 구글 SMTP 사용자 이름 값으로 설정

Minjae-An committed:
feat: 임시 비밀번호 발급 비즈니스 로직 추가(#35)

임시 비밀번호는 다음 과정을 거친다.
1. 요청 이메일 기반 사용자 조회
2. 임시 비밀번호 생성
3. 사용자 비밀번호, 임시 비밀번호로 변경
4. 임시 비밀번호 발급 메일 전송
임시 비밀번호를 암호화하는 로직은 추후 구현 예정

Minjae-An committed:
feat: 임시 비밃번호 발급 비즈니스 로직 트랜잭션 정의 추가(#35)

Minjae-An committed:
feat: 임시 비밀번호 발급 요청 DTO 수정(#35)

- 불필요한 기존 생성자 삭제, 기본 생성자 추가
- 비밀번호 형식 제약 조건 추가
- 코드 정렬

Minjae-An committed:
feat: 임시 비밀번호 발급 API 구현(#35)

- 임시 비밀번호 발급 API 구현
- 명세 속성 수정
- 코드 정렬

Minjae-An committed:
Merge branch 'dev' into feat/#35-send-temporary-password

# Conflicts:
#	be/src/main/java/yeonba/be/arrow/service/ArrowService.java
#	be/src/main/java/yeonba/be/user/controller/UserController.java
#	be/src/main/java/yeonba/be/user/entity/User.java

Minjae-An committed:
Merge branch 'dev' into feat/#35-send-temporary-password

# Conflicts:
#	be/src/main/java/yeonba/be/user/entity/User.java

Minjae-An committed:
Merge branch 'dev' into feat/#35-send-temporary-password

Minjae-An committed:
refactor: 병합에 따른 코드 재배치(#35)

dev 브랜치 작업 내역 병합에 따른 코드 재배치

Minjae-An committed:
refactor: 병합에 따른 코드 재배치(#35)

dev 브랜치 작업 내역 병합에 따른 코드 재배치

Minjae-An committed:
chore: sms 서비스 sdk 의존성 추가(#28)

coolsms 서비스에서 제공하는 java sdk 의존성 추가

Minjae-An committed:
chore: sms 서비스 관련 설정 추가(#28)

sms 서비스를 이용하기 위해 필요한 설정들 추가

Minjae-An committed:
feat: sms 서비스 이용을 위한 설정 정의(#28)

- coolsms에서는 메시지 전송을 수행할 수 있는 DefaultMessageService 제공
- config에서 application.yml을 통해 값을 주입 받아 DefaultMessageService 빈 등록

Minjae-An committed:
feat: 인증 코드 생성 로직 추가(#28)

- apache.common.lang3 라이브러리의 RandomStringUtils 이용
- 영어 대소문자, 숫자로 구성된 6자리 랜덤 인증 코드를 생성
- 영어 대소문자 52개(26+26), 숫자 10개로 총 62개의 선택 가능 문자 존재
- 생성 가능한 인증 코드 수, 약 56억 8천만 개(62^6)로 보안성을 확보하였음

Minjae-An committed:
feat: 문자 전송 로직 추가(#28)

Minjae-An committed:
feat: 전화번호 기반 사용자 존재 여부 확인 로직 정의(#28)

Minjae-An committed:
feat: 전화번호 기반 사용자 존재 여부 확인 로직 추가(#28)

Minjae-An committed:
feat: 전화번호 형식 제약 조건 추가(#28)

010으로 시작하며 0~9까지의 숫자로 이뤄진 11자리 문자열만 허용토록 제약 조건 추가

Minjae-An committed:
chore: application.yml 수정(#28)

sms.provider 속성도 application.properties 통해 관리하도록 수정

Minjae-An committed:
feat: 인증 코드 sms 전송 비즈니스 로직 추가(#28)

- 요청된 전화번호를 통해 사용자 존재 확인
- 사용자가 존재할 경우만 sms로 인증 코드 전송

Minjae-An committed:
feat: 전화번호 인증 코드 전송 API 구현(#28)

Minjae-An committed:
chore: spring data redis 의존성 추가(#28)

Minjae-An committed:
chore: redis 관련 설정 추가(#28)

Minjae-An committed:
feat: redis 관련 설정 정의(#28)

- redis 커넥션 팩토리 빈 등록
- CRUD 연산 지원을 위한 RedisTemplate 빈 등록

Minjae-An committed:
feat: redis 키-값 저장,조회,삭제 로직 추가(#28)

- redis를 이용한 저장, 삭제 기능 지원 유틸 클래스 정의
- 만료 시간을 분 단위로 설정하는 저장 로직 구현
- 기본 조회 로직 구현, Optional 리턴 타입을 통해 null 처리 지원
- 기본 삭제 로직 구현

Minjae-An committed:
feat: 인증 코드 sms 전송 비즈니스 로직 수정(#28)

- 인증 코드 sms 전송 이전 redis에 번호를 키로 코드를 저장하는 로직 추가
- 새 인증 코드 발급 이전 기존 발급 내역 삭제

Minjae-An committed:
feat: 인증 코드 정규식 정의(#28)

영어 대소문자, 숫자로 이뤄진 6자리 문자열만 허용하는 정규식 enum 정의

Minjae-An committed:
feat: 아이디 찾기 예외 enum 정의(#28)

- 인증 코드 내역이 존재하지 않는 경우(인증 코드 만료도 포함)
- 인증 코드가 일치하지 않는 경우

Minjae-An committed:
feat: 전화번호 기반 사용자 조회 기능 정의(#28)

Minjae-An committed:
feat: 전화번호 기반 사용자 조회 로직 추가(#28)

Minjae-An committed:
refactor: 코드 정렬(#28)

Minjae-An committed:
feat: 아이디 찾기 요청 DTO 수정(#28)

- 인증 번호를 발급한 전화번호 필드 추가
- 인증 코드 필드 명세 속성 수정
- 인증 코드 정규식 제약 조건 추가
- 불필요한 기존 생성자, 기본 생성자로 대체

Minjae-An committed:
feat: 아이디 찾기 비즈니스 로직 추가(#28)

아이디 찾기 요청은 인증 코드 발급 전화번호, 인증 코드를 포함한다.
아이디 찾기는 다음 과정을 거쳐 이뤄진다.
1. 발급 전화번호를 기반으로 인증 코드 redis에서 조회
2. 인증 코드 일치 확인
3. 전화번호 기반 사용자 조회
4. 확인한 인증 코드 삭제
5. 찾은 이메일을 응답으로 반환

Minjae-An committed:
feat: 아이디 찾기 API 구현(#28)

Minjae-An committed:
feat: 전화번호 정규식 정의(#28)

Minjae-An committed:
feat: 회원가입 요청 DTO 수정(#8)

- 요구사항 변경에 따라 필드들 삭제 및 추가
- 필드들 API 명세
- 문자열 필드들 정규식 제약 조건 추가
- 불필요한 생성자 삭제, 기본 생성자로 대체

Minjae-An committed:
Merge branch 'feat/#28-verification-code-to-phone' into feat/#8-join-api

# Conflicts:
#	be/src/main/java/yeonba/be/mypage/dto/response/UserProfileDetailResponse.java
#	be/src/main/java/yeonba/be/user/dto/response/UserProfileResponse.java
#	be/src/main/java/yeonba/be/user/entity/User.java

Minjae-An committed:
refactor: PasswordEncryptor 최상위 util 패키지로 이동(#8)

- 비밀번호 암호화 기능을 전역적으로 사용하기에 패키지 경로 변경
- 사용 위치 수정

Minjae-An committed:
refactor: 잘못된 들여쓰기 수정을 위해 코드 정렬(#8)

Minjae-An committed:
feat: 이름 기반 동물상 조회 로직 정의(#8)

Minjae-An committed:
feat: 이름 기반 닮은 동물상 조회 로직 구현(#8)

Minjae-An committed:
feat: 이름 기반 지역 조회 로직 정의(#8)

Minjae-An committed:
feat: 이름 기반 지역 조회 로직 구현(#8)

Minjae-An committed:
feat: 분류 기반 음역대 조회 로직 정의(#8)

Minjae-An committed:
feat: 분류 기반 음역대 조회 로직 구현(#8)

Minjae-An committed:
feat: 사용자 선호 조건 repository 정의(#8)

Minjae-An committed:
feat: 사용자 선호 조건 엔티티 저장 로직 구현(#8)

Minjae-An committed:
feat: 사용자 선호조건 엔티티 구성(#8)

이성 추천, 다른 사용자 검색시 사용되는 사용자 선호조건 엔티티 정의
다음 필드들을 포함한다.
- 선호하는 나이 하한, 상한
- 선호하는 키 하한, 상한
- 선호하는 MBTI
- 선호하는 체형
- 선호하는 음역대
- 선호하는 지역
- 선호하는 동물상
- 선호 조건 소유 사용자

Minjae-An committed:
feat: 사용자 엔티티 저장 로직 구성(#8)

Minjae-An committed:
feat: 프로필 사진 repository 정의(#8)

Minjae-An committed:
feat: 프로필 사진 엔티티 저장 로직 구성(#8)

Minjae-An committed:
feat: 회원가입 관련 예외 enum 정의(#8)

다음 예외 enum 들을 정의
- 비밀번호와 비밀번호 확인 값 불일치 상황
- 존재하지 않는 음역대
- 존재하지 않는 동물상
- 존재하지 않는 지역

Minjae-An committed:
refactor: 분류 기반 음역대 조회 로직 수정(#8)

- 잘못된 인덴트 수정, 코드 정렬
- 회원가입 관련 예외 enum을 사용토록 로직 수정

Minjae-An committed:
refactor: 이름 기반 닮은 동물상 조회 기능 수정(#8)

- 잘못된 들여쓰기 수정, 코드 정렬
- 회원가입 관련 예외 enum을 이용하도록 수정

Minjae-An committed:
refactor: 이름 기반 지역 조회 로직 수정(#8)

- 잘못된 들여쓰기, 코드 정렬 수정
- 회원가입 관련 예외 enum 사용토록 수정

Minjae-An committed:
refactor: 잘못된 들여쓰기, 코드 정렬로 수정(#8)

Minjae-An committed:
refactor: 회원가입 요청 DTO 수정(#8)

- 기존 생성자 삭제, @ModelAttribute 사용을 위한 생성자 추가
- 잘못된 들여쓰기 수정을 위한 코드 정렬 수정

Minjae-An committed:
feat: 회원가입 응답 DTO 구성(#8)

회원가입시 바로 서비스를 이용할 수 있도록 access token과
refresh token을 바로 지급하는 형태로 구성

Minjae-An committed:
feat: 성별 enum 구성(#8)

- 성별 enum 정의
- 문자열 성별 값을 enum으로 치환할 수 있는 편의 메서드 구현

Minjae-An committed:
feat: 비밀번호 암호화시 사용되는 salt 생성 로직 구현(#8)

Minjae-An committed:
refactor: 잘못된 들여쓰기 수정, 코드 정렬(#8)

Minjae-An committed:
feat: 회원가입 관련 로직 추가(#8)

- 사용자 엔티티 생성 및 저장 로직
- 프로필 사진 엔티티 생성 및 저장 로직
- 사용자 선호조건 엔티티 생성 및 저장 로직

Minjae-An committed:
feat: 회원가입 비즈니스 로직 초안 구성(#8)

회원가입 비즈니스 로직은 다음 과정을 거쳐 이뤄진다.
1. 사용자 엔티티 생성 및 저장
2. 프로필 사진 엔티티 생성 및 저장
3. 사용자 선호조건 엔티티 생성 및 저장
4. access token, refresh token 발급

Minjae-An committed:
feat: 회원가입 API 초안 구현 및 코드 정렬(#8)

- 회원가입 API 구현
- 잘못된 들여쓰기 수정을 위한 코드 정렬

Minjae-An committed:
refactor: 분류 기반 음역대 조회 로직 이름 수정, 코드 정렬(#8)

Minjae-An committed:
feat: 프로필 사진 엔티티 저장 로직 수정(#8)

무조건 일괄적 List 형태로 저장되는 특성을 고려해 List 파라미터로 받아
일괄 저장하도록 로직 수정

Minjae-An committed:
feat: S3 이용 프로필 사진 업로드 로직 추가(#8)

프로필 사진 업로드는 다음 과정을 거쳐 이뤄진다.
1. 사진 파일들 확장자 검증
2. 파일 식별을 위한 키 생성, profilephoto/{userId}-{photo idx} 형식
3. 사진 파일 업로드 요청 생성 및 업로드 수행

Minjae-An committed:
feat: 프로필 사진 엔티티 저장 로직에 S3 업로드 로직 추가(#8)

Minjae-An committed:
feat: 프로필 사진 컬렉션 필드 초기화 & 연관관계 설정 로직 추가(#8)

- NPE 방지를 위한 컬렉션 필드 초기화 설정
- 순수한 객체 관계를 고려한 프로필 사진 엔티티 연관관계 설정 로직 구성

Minjae-An committed:
feat: 이미 사용 중인 닉네임, 이메일 확인 로직 정의(#8)

- 이미 사용중인 닉네임 확인 로직 정의
- 이미 사용중인 이메일 확인 로직 정의

Minjae-An committed:
feat: 이미 사용 중인 닉네임, 이메일 확인 로직 추가(#8)

Minjae-An committed:
feat: 랜덤 salt 생성 로직 수정(#8)

- salt의 길이 고정 32비트(4바이트)로 변경
- 32비트 길이 salt 만으로 대부분의 보안적 위협 충분히 커버 가능
- base64 인코딩에 따라 로직에서 생성되는 salt 문자열의 길이는 8

Minjae-An committed:
feat: 이미 사용 중인 이메일/닉네임 예외 추가(#8)

- 이미 사용 중인 이메일 예외 enum 정의
- 이미 사용 중인 닉네임 예외 enum 정의

Minjae-An committed:
feat: s3 업로드된 프로필 사진 삭제 로직 추가(#8)

Minjae-An committed:
feat: 회원가입 관련 로직 수정(#8)

- 사용자 엔티티 저장 로직에 이미 사용 중인 이메일/닉네임 검증 로직 추가
- 프로필 사진 저장 로직에 사용자 엔티티와의 연관관계 설정 로직 추가

Minjae-An committed:
feat: 프로필 사진 삭제 이벤트 정의(#8)

- 회원가입 과정중 트랜잭션 롤백 발생시 업로드된 프로필 사진 삭제를 위해 정의
- 업로드된 프로필 사진 엔티티를 필드를 통해 전달

Minjae-An committed:
feat: 프로필 사진 삭제 이벤트 리스너 정의(#8)

- 회원가입 과정중 트랜잭션 롤백 발생시 프로필 사진 삭제 이벤트 발생
- 이벤트를 통해 삭제 대상인 프로필 사진 엔티티들을 전달 받음
- S3 서비스를 통해 해당 엔티티들의 삭제 작업 진행

Minjae-An committed:
feat: 회원가입 로직 롤백시 업로드된 프로필 사진 삭제 이벤트 발행 로직 추가(#8)

회원가입 도중 예외 상황 등으로 인해 트랜잭션 롤백시 이미 s3에 업로드된 프로필
사진들을 삭제하기 위한 이벤트 발행 로직 추가

Minjae-An committed:
chore: jjwt 의존성 추가(#8)

jwt 활용을 위해 사용되는 jjwt 라이브러리 의존성 추가

Minjae-An committed:
feat: access/refresh token 발급 및 검증 로직 추가(#8)

- jwt 시크릿 키 값의 경우 무작위 64바이트 바이너리를 16진수로 인코딩한 문자열 사용
- jwt 포맷의 access/refresh token 발급 로직 추가
- 만료 시간을 계산하는 공통 로직, 별도의 메서드로 추출
- 사용자 엔티티를 기반으로 jwt를 생성하는 공통 로직, 별도의 메서드로 추출
- access token의 보편적인 존속기간은 수 시간가량 따라서 8시간으로 설정
- refresh token의 보편적인 존속기간은 수 일에서 수 개월 내외 따라서 10일로 설정
- jwt의 일반적인 관행에 따라 사용자 이메일을 subject로 설정
- 사용자 엔티티의 PK를 클레임으로 설정하여 인가시 사용할 수 있도록 구성
- 검증 로직에선 시그니처의 유효성과 JWT 만료 여부를 검증

Minjae-An committed:
feat: refresh token 필드 및 업데이트 로직 추가(#8)

- 사용자의 refresh token을 저장할 필드 추가
- 사용자 refresh token 업데이트 로직 추가

Minjae-An committed:
feat: 회원가입 로직, access/refresh token 발급 로직 추가(#8)

- 회원가입 정상 완료의 결과로 access token과 refresh token 제공
- 회원가입 비즈니스 로직에 access token, refresh token 발급 로직 추가
- 새 refresh token 발급시 사용자 refresh token 업데이트 수행

Minjae-An committed:
refactor: 잘못된 주석 들여쓰기 수정(#8)

Minjae-An committed:
refactor: 문자열 값 Geneder enum 변환 메서드 이름 수정(#8)

Minjae-An committed:
refactor: 회원가입 비즈니스 로직 코드 정렬(#8)

Minjae-An committed:
refactor: Gender 문자열 값 변환 메서드 사용 위치 수정 및 코드 정렬(#8)

Minjae-An committed:
feat: salt 길이 32바이트로 변경(#8)

보안적으로 안전한 32자로 salt 길이 변경

Minjae-An committed:
Merge branch 'dev' into feat/#8-join-api

# Conflicts:
#	be/build.gradle
#	be/src/main/java/yeonba/be/config/MailConfig.java
#	be/src/main/java/yeonba/be/login/controller/LoginController.java
#	be/src/main/java/yeonba/be/login/dto/request/UserPasswordInquiryRequest.java
#	be/src/main/java/yeonba/be/login/service/LoginService.java
#	be/src/main/java/yeonba/be/user/entity/User.java
#	be/src/main/java/yeonba/be/user/repository/UserQuery.java
#	be/src/main/java/yeonba/be/user/repository/UserRepository.java
#	be/src/main/java/yeonba/be/util/EmailService.java
#	be/src/main/java/yeonba/be/util/TemporaryPasswordGenerator.java
#	be/src/main/resources/application.yml

Minjae-An committed:
refactor: 병합으로 코드 수정(#8)

Minjae-An committed:
feat: 임시 비밀번호 발급 로직에 비밀번호 암호화 로직 추가(#8)

Minjae-An committed:
feat: 비밀번호 불일치 예외 정의(#50)

Minjae-An committed:
feat: 로그인 요청 dto, 검증 조건 추가(#50)

dto 필드들에 다음 검증 조건들을 추가하였다.
- 이메일 공백값, null 값 검증 조건
- 이메일 정규식 검증 조건
- 비밀번호 공배값, null 값 검증 조건
- 비밀번호 정규식 검증 조건

Minjae-An committed:
refactor: 잘못된 들여쓰기 수정, 코드 정렬(#50)

Minjae-An committed:
feat: 로그인 비즈니스 로직 추가(#50)

로직은 다음 과정을 거쳐 수행된다.
1. 이메일로 사용자 조회
2. 비밀번호 요청값 암호화 및 저장된 비밀번호와 일치 확인
3. access token, refresh token 발급
4. 사용자 refresh token 업데이트
5. 응답 반환

Minjae-An committed:
feat: 로그인 API 구현(#50)

Minjae-An committed:
refactor: access token 재발급 요청 dto 수정(#50)

- refresh token 필드 예시 값 수정
- 공백/null 값 검증 조건 추가
- 불필요한 생성자 삭제

Minjae-An committed:
feat: 커스텀 Jwt 예외 정의(#50)

이하 3가지 Jwt 관련 예외를 커스텀 처리하기 위해 예외 정의
- 유효하지 않은 JWT 형식
- 유효하지 않은 시그니처
- 만료된 JWT

Minjae-An committed:
feat: refresh token 불일치 예외 정의(#50)

Minjae-An committed:
feat: jwt 유틸 로직 추가(#50)

- access token 생성 로직
- refresh token 생성 로직
- 만료 일자 생성 로직
- 사용자 jwt 생성 로직
- jwt 검증 로직

Minjae-An committed:
feat: access token 재발급 비즈니스 로직 추가(#50)

access token 재발급은 다음 과정을 거쳐 이뤄진다.
1. refresh token에서 userId 파싱
2. id 기반 사용자 조회
3. refresh token 입력값과 사용자 refresh token 일치 확인
4. access token 재발급

Minjae-An committed:
feat: access token 재발급 API 구현(#50)

Minjae-An committed:
feat: Bearer token 검증 정규식 정의(#50)

Minjae-An committed:
feat: access token 인가 관련 예외 정의(#50)

- access token이 존재하지 않는 경우
- bearer token 형식이 유효하지 않은 경우

Minjae-An committed:
feat: access token 이용, 인가 로직 추가(#50)

access token을 이용한 인가는 이하 과정을 거쳐 이뤄진다.
1. Authorization 헤더 값이 존재하는 지 확인
2. Bearer 토큰 형식이 맞는 지 확인
3. jwt 검증(형식, 시그니처 유효성, 만료 여부)
4. access token 파싱, userId 파싱
5. userId request attribute 설정

Minjae-An committed:
Merge branch 'dev' into feat/#50-login-and-access-token-refresh

# Conflicts:
#	be/build.gradle
#	be/src/main/java/yeonba/be/config/SmsConfig.java
#	be/src/main/java/yeonba/be/exception/LoginException.java
#	be/src/main/java/yeonba/be/login/controller/LoginController.java
#	be/src/main/java/yeonba/be/login/dto/request/UserIdInquiryRequest.java
#	be/src/main/java/yeonba/be/login/dto/request/UserPhoneNumberVerifyRequest.java
#	be/src/main/java/yeonba/be/login/dto/response/UserEmailInquiryResponse.java
#	be/src/main/java/yeonba/be/login/service/LoginService.java
#	be/src/main/java/yeonba/be/mypage/service/MyPageService.java
#	be/src/main/java/yeonba/be/user/entity/User.java
#	be/src/main/java/yeonba/be/user/repository/UserQuery.java
#	be/src/main/java/yeonba/be/user/repository/UserRepository.java
#	be/src/main/java/yeonba/be/util/ServiceRegex.java
#	be/src/main/java/yeonba/be/util/SmsService.java
#	be/src/main/java/yeonba/be/util/VerificationCodeGenerator.java

Minjae-An committed:
refactor: 병합으로 인한 코드 수정(#50)

dev 브랜치 작업내역 병합으로 인한 코드 수정

Minjae-An committed:
Merge branch 'dev' into feat/#50-login-and-access-token-refresh

# Conflicts:
#	be/build.gradle
#	be/src/main/java/yeonba/be/exception/JoinException.java
#	be/src/main/java/yeonba/be/exception/LoginException.java
#	be/src/main/java/yeonba/be/login/controller/LoginController.java
#	be/src/main/java/yeonba/be/login/service/LoginService.java

Minjae-An committed:
refactor: 병합에 따른 코드 수정(#50)

dev 브랜치 작업내역 병합에 따른 코드 수정

Minjae-An committed:
Merge branch 'dev' into feat/#50-login-and-access-token-refresh

# Conflicts:
#	be/src/main/java/yeonba/be/login/dto/request/UserJoinRequest.java
#	be/src/main/java/yeonba/be/login/dto/response/UserJoinResponse.java
#	be/src/main/java/yeonba/be/login/service/LoginService.java
#	be/src/main/java/yeonba/be/user/dto/response/UserProfileResponse.java
#	be/src/main/java/yeonba/be/user/entity/ProfilePhoto.java
#	be/src/main/java/yeonba/be/user/entity/UserPreference.java
#	be/src/main/java/yeonba/be/user/enums/Gender.java
#	be/src/main/java/yeonba/be/user/repository/UserCommand.java
#	be/src/main/java/yeonba/be/user/repository/UserQuery.java
#	be/src/main/java/yeonba/be/user/repository/animal/AnimalQuery.java
#	be/src/main/java/yeonba/be/user/repository/animal/AnimalRepository.java
#	be/src/main/java/yeonba/be/user/repository/area/AreaQuery.java
#	be/src/main/java/yeonba/be/user/repository/area/AreaRepository.java
#	be/src/main/java/yeonba/be/user/repository/profilephoto/ProfilePhotoCommand.java
#	be/src/main/java/yeonba/be/user/repository/userpreference/UserPreferenceCommand.java
#	be/src/main/java/yeonba/be/user/repository/vocalrange/VocalRangeQuery.java
#	be/src/main/java/yeonba/be/user/repository/vocalrange/VocalRangeRepository.java
#	be/src/main/java/yeonba/be/user/service/JoinService.java
#	be/src/main/java/yeonba/be/user/service/UserService.java
#	be/src/main/java/yeonba/be/util/JwtUtil.java
#	be/src/main/java/yeonba/be/util/S3Service.java
#	be/src/main/java/yeonba/be/util/SaltGenerator.java

Minjae-An committed:
refactor: 불필요한 클래스 삭제(#50)

Minjae-An committed:
refactor: 병합으로 인한 코드 수정(#50)

Minjae-An committed:
refactor: access token not exist 예외 메시지 수정(#50)

Minjae-An committed:
refactor: jwt 검증 로직 가시성 private 설정(#50)

Minjae-An committed:
refactor: 주석 및 코드 정렬 수정(#50)

Minjae-An committed:
refactor: 병합으로 인한 수정(#50)

- 병합 과정에서 잘못 반영된 수정 사항 반영
- ServiceRegex -> GlobalValidationRegex로 이름 수정
- 그에 따른 사용 위치 수정

Minjae-An commented:
JWT 예외 관련 수정사항으로 인해 close

Minjae-An committed:
refactor: JWT 예외, UtilException으로 처리토록 로직 수정(#50)

- 어떤 예외 상황이든 JWT가 유효하지 않다는 것은 똑같다
- UtilException.INVALID_JWT를 통해 일괄 예외 처리토록 로직 수정

Minjae-An commented:
출석 체크 비즈니스 로직 수정사항으로 인해 close

Minjae-An committed:
refactor: 출석 체크 여부 확인 메서드 수정(#50)

로직에서 직접 예외를 발생시키는 형태에서 boolean 반환 형태로 변경

Minjae-An committed:
refactor: 화살 보낸 사용자, 받은 사용자 필드 이름 변경(#50)

- 보다 직관적으로 나타내도록 필드 이름 변경
- 사용 위치 변경

Minjae-An committed:
refactor: 화살 내역 엔티티 필드명 변경에 따른 수정(#50)

- sent_user -> sender
- received_user -> receiver
위 변경에 따라 연관 로직 수정

Minjae-An committed:
refactor: 사용자 출석 체크 여부 확인 로직 변경에 따른 수정(#50)

Minjae-An committed:
feat: 존재하지 않는 refresh token 예외 정의(#50)

Minjae-An committed:
refactor: 잘못된 파라미터명 수정(#50)

Minjae-An committed:
refactor: access token 재발급 로직 파라미터 타입 변경(#50)

Minjae-An committed:
refactor: 로그인, access token 재발급 API 로직 변경(#50)

- 기존 로그인 로직에서는 response body를 통해 refresh token 제공
- 이 형태는 탈취 가능성이 높음, 따라서 적절한 속성을 가진 쿠키를 통해
  제공토록 변경
- HttpOnly=true을 통해 XSS 공격을 방지
- SameSite=true 속성을 통해 CSRF 공격을 방지
- secure=true 설정을 통해 https 환경에서만 쿠키를 전송토록 설정
- access token 재발급 요청시에만 refresh token cookie를
  보내도록 path 설정 추가
- refresh token을 제공하는 방식이 달라짐에 따라 access token 재발급
  API에서 refresh token을 쿠키에서 탐색해 service에 넘기도록 변경

Minjae-An committed:
refactor: access token 생성 로직 수정(#50)

Bearer 토큰 형식으로 제공하도록 생성 로직 수정

Minjae-An committed:
refactor: dto 이름 변경(#50)

access token을 응답하는 용도로 사용된다는 것을 나타낼 수 있게 이름 변경

Minjae-An committed:
refactor: access token 재발급 로직 수정(#50)

반환 dto 타입 이름 변경에 따른 수정

Minjae-An committed:
refactor: 불필요한 주석 삭제(#50)

Minjae-An committed:
refactor: refresh token 발급 관련 로직 수정(#50)

- refresh token을 쿠키에 설정하는 공통 로직 별도의 메서드 추출
- 회원가입시 access token과 refresh token 같이 발급하여 제공
- 회원가입 로직에 응답에 refresh token cookie 포함하는 로직 추가
- 메서드 추출에 따른 기존 로직 리팩토링

Minjae-An committed:
feat: 로그인 결과 전달을 위한 dto 정의(#50)

Minjae-An committed:
refactor: 사용자 로그인 응답 DTO 수정(#50)

- refresh token을 쿠키를 통해 전달함에 따라 관련 필드 삭제
- 사용자 휴면 여부 필드 추가

Minjae-An committed:
refactor: 로그인 로직 반환 타입 변경(#50)

access token, refresh token, 사용자 휴면 여부를 함께
controller에 전달할 수 있는 UserLoginResult 타입으로 변경

Minjae-An committed:
refactor: 로그인 API 로직 수정(#50)

- refresh token을 클라이언트가 쿠키를 통해 서버에 전달토록
  controller에서 응답 헤더의 Set-Cookie를 통해 refresh token을
  전달
- 한편, 로그인시 access token과 사용자 휴면 여부가 response body를
  통해 전달되어야 함
- service에서 필요한 모든 데이터를 UserLoginResult 객체를 통해
  전달받고 controller에서 적절히 가공하여 응답하도록 로직 수정

acceptor-gyu commented on be/src/main/java/yeonba/be/login/dto/request/UserLoginRequest.java:
```diff
@@ -1,26 +1,34 @@
-      example = "mj3242@naver.com"
-  )
-  private String email;
+	@Schema(
+		type = "string",
+		description = "이메일",
+		example = "mj3242@naver.com")
+	@Pattern(
+		regexp = "[a-zA-Z0-9_!#$%&’*+/=?`{|}~^.-]+@[a-zA-Z0-9.-]+$",
+		message = "유효하지 않은 이메일 형식입니다.")
+	@NotBlank(message = "이메일은 반드시 입력되어야 합니다.")
```
tap 사용하지 않기로 했는데 사용된 것같습니다. 확인부탁드립니다.

acceptor-gyu commented on be/src/main/java/yeonba/be/login/service/LoginService.java:
```diff
@@ -153,4 +158,47 @@ public void deleteExpiredVerificationCodes() {
+            throw new GeneralException(LoginException.PASSWORD_NOT_MATCH);
+        }
+
+        // access token, refresh token 발급
+        Date issuedAt = new Date();
+        String accessToken = jwtUtil.generateAccessToken(user, issuedAt);
+        String refreshToken = jwtUtil.generateRefreshToken(user, issuedAt);
```
issuedAt이라는 변수명은 수정하기로 얘기된 거 아니었나요?

acceptor-gyu reviewed (changes requested):
형식 및 사소한 실수나 이전에 얘기됐지만 수정되지 않은 부분들 수정하시면 다시 확인하겠습니다.

acceptor-gyu commented:
단순히 기능 구현만 생각을 하지 마시고 리뷰하는 사람을 생각해서 본인의 PR을 다시 한 번 검토하는 시간을 가져보는 것도 좋을 것같습니다.

Minjae-An committed:
refactor: 사용하지 않는 refresh token 요청 dto 삭제(#50)

Minjae-An committed:
refactor: 잘못된 들여쓰기 수정(#50)

Minjae-An committed:
refactor: 잘못된 병합으로 인한 코드 수정(#50)

Minjae-An committed:
refactor: issuedAt 변수명, generatedAt으로 일괄 수정(#50)

Minjae-An commented:
말씀하신 대로 너무 기능 구현에 급급하였던 것 같습니다. PR을 올릴 때 더 세심히 검토하겠습니다. 죄송합니다.

Minjae-An committed:
refactor: 불필요한 주석 제거(#50)

Minjae-An committed:
refactor: 회원 가입 응답 dto 수정(#50)

- refresh token을 응답 dto가 아닌 response header의 Set-Cookie를 통해 전달
- 이에 따라 response body에 사용하던 dto를 service에서 controller로
  로직 결과를 전달하는 용도로만 사용하게 변경
- 클래스 이름, 패키지 경로 적절히 변경

Minjae-An committed:
refactor: jwt 필드 API 명세 속성 일괄 수정(#50)

보다 간단한 형태로 jwt 필드라는 것을 나타내도록 일괄 수정

Minjae-An committed:
refactor: 회원가입 로직 반환 타입 변경(#50)

Minjae-An committed:
refactor: 회원 가입 API, 로직 수정(#50)

회원가입 비즈니스 로직 반환 타입 변경에 따른 수정

Minjae-An commented:
소셜 로그인으로의 변경에 따른 close
